### PR TITLE
Response aligning of get_info & get_account

### PR DESCRIPTION
**Pull request by psy2848048**
## Background

- Many responses are changed
- Precision of big number is not kept
- No mock test of API

## Summary

- Added http mock test & mock server
- Added Int & Decimal type (must be duplicates of Int64 & Float64, my implementation will replace or be replaced them by review)
- Aligned the response of `get_info` & `get_account`
- Added Go 1.18, 1.19 in CI test

## Note

- Kind of code reading & kickoff contribution
- Other aligns will be following

## Checklist

- [x] Backward compatible?
- [x] Test enough in your local environment?
- [x] Add related test cases?

### CODE REVIEW

File: api_test.go [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+	}
+
+	expectedJSON := mockserver.OpenFile("chain_get_account.json")
+
+	assert.JSONEq(t, expectedJSON, string(actualJSON))
+}
+
+func TestAPIGetInfo(t *testing.T) {
+	info, err := api.GetInfo(context.Background())
+	if err != nil {
```
Comment (maoueh) [inline api_test.go]: Just keep the `assert.` call below, also, we usually use `require.NoError` for errors so that the test stops right there, there is usually no good reason to continue the test if there is an error because there is it's almost sure that all other assertions will fails.

File: api_test.go [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+		panic(fmt.Errorf("get info: %w", err))
+	}
+
+	assert.NoError(t, err)
+
+	actualJSON, err := json.Marshal(info)
+	if err != nil {
+		panic(err)
+	}
```
Comment (maoueh) [inline api_test.go]:
```suggestion
    require.NoError(t, err)
```

File: responses.go [hunk 3]
```diff
@@ -222,9 +239,18 @@ type AccountResp struct {
 	CPULimit               AccountResourceLimit `json:"cpu_limit"`
 	Permissions            []Permission         `json:"permissions"`
 	TotalResources         TotalResources       `json:"total_resources"`
-	SelfDelegatedBandwidth DelegatedBandwidth   `json:"self_delegated_bandwidth"`
+	SelfDelegatedBandwidth *DelegatedBandwidth  `json:"self_delegated_bandwidth"`
 	RefundRequest          *RefundRequest       `json:"refund_request"`
 	VoterInfo              VoterInfo            `json:"voter_info"`
+
+	// added 2022-11-02
```
Comment (maoueh) [inline responses.go]: We should actually write down in which version of EOSIO/LEAP this was actually added/changed rather than the date, as it can easily be found from a `git blame`.

Also, leaving a blank line before the fields so the comment is not attached to the field directly but is more informational in the code.

Please change it throughout the PR.

File: responses.go [hunk 4]
```diff
@@ -222,9 +239,18 @@ type AccountResp struct {
...
+	HeadBlockNum           uint64               `json:"head_block_num"`
+	HeadBlockTime          BlockTimestamp       `json:"head_block_time"`
+	RexInfo                *RexInfo             `json:"rex_info,omitempty"`
+	SubjectiveSpuBillLimit AccountResourceLimit `json:"subjective_cpu_bill_limit"`
+
+	// TODO: need to set the exact type
```
Comment (maoueh) [inline responses.go]: Can I help somehow with this TODO?

File: api_test.go [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+	assert.JSONEq(t, expectedJSON, string(actualJSON))
+}
+
+func TestMain(m *testing.M) {
+	setUp()
+	code := m.Run()
+	tearDown()
+
+	os.Exit(code)
+}
```
Comment (maoueh) [inline api_test.go:47]: I've been doing Go for the last 5 years, first time I saw that (or at least that I remember) amazing!

File: testdata/mock_server/server.go [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+
+func OpenFile(filename string) string {
+	path := []string{"./testdata/mock_server", filename}
+	body, err := os.ReadFile(strings.Join(path, "/"))
+	if err != nil {
+		panic(err)
+	}
+
+	return string(body)
+}
```
Comment (maoueh) [inline testdata/mock_server/server.go]: The hard-coded path currently works because `OpenFile` is used within a test that is at the root of the project. Using it from another sub-package would fail. This is because Go tests are run in their package's directory.

It would be better to pass the full path to it. But this needs to be passed down from `CreateAndActivateRestMockServer`.

To reduce the having to pass `testdata/mock_server` everywhere, I suggest `CreateAndActivateRestMockServer` receives where is located `rootPackageDirectory`.
```suggestion
func OpenFile(rootPackageDirectory string, filename string) string {
	body, err := os.ReadFile(filepath.Join(rootPackageDirectory, "testdata/mock_server", filename))
	if err != nil {
		panic(err)
	}

	return string(body)
}
```

And `CreateAndActivateRestMockServer` would be call like `CreateAndActivateRestMockServer(".")` and `mockserver.OpenFile(".", <filename>)`.

Also, do not use `strings.Join` to join path segment, use `filepath.Join` instead which deals with platform differences.

File: responses.go [hunk 7]
```diff
@@ -211,8 +228,8 @@ type ProducerChange struct {
 type AccountResp struct {
 	AccountName            AccountName          `json:"account_name"`
 	Privileged             bool                 `json:"privileged"`
-	LastCodeUpdate         JSONTime             `json:"last_code_update"`
-	Created                JSONTime             `json:"created"`
+	LastCodeUpdate         BlockTimestamp       `json:"last_code_update"` // modified 2022-11-02
+	Created                BlockTimestamp       `json:"created"`          // modified 2022-11-02
```
Comment (maoueh) [inline responses.go]: What the reasoning of those changes?

File: responses.go [hunk 8]
```diff
@@ -222,9 +239,18 @@ type AccountResp struct {
 	CPULimit               AccountResourceLimit `json:"cpu_limit"`
 	Permissions            []Permission         `json:"permissions"`
 	TotalResources         TotalResources       `json:"total_resources"`
-	SelfDelegatedBandwidth DelegatedBandwidth   `json:"self_delegated_bandwidth"`
+	SelfDelegatedBandwidth *DelegatedBandwidth  `json:"self_delegated_bandwidth"`
```
Comment (maoueh) [inline responses.go]: And this one? So that it can be `null`?

File: responses.go [hunk 9]
```diff
@@ -42,6 +51,14 @@ type InfoResp struct {
 	BlockCPULimit        Int64  `json:"block_cpu_limit"`
 	BlockNetLimit        Int64  `json:"block_net_limit"`
 	ServerVersionString  string `json:"server_version_string"`
+
+	EarliestAvailableBlockNum uint64         `json:"earliest_available_block_num"`
```
Comment (maoueh) [inline responses.go]: Added `// Added in <version>` comment

File: types.go [hunk 10]
```diff
@@ -610,10 +616,32 @@ func (a Asset) MarshalJSON() (data []byte, err error) {
 	return json.Marshal(a.String())
 }
 
+// added 2022-11-02
```
Comment (maoueh) [inline types.go]: Use standard Go comment structure here which should start with the struct's name so like:

```
// RexInfo was added on version <>
```

File: types.go [hunk 11]
```diff
@@ -610,10 +616,32 @@ func (a Asset) MarshalJSON() (data []byte, err error) {
...
+	Owner      AccountName `json:"owner"`
+	VoteStake  Asset       `json:"vote_stake"`
+	RexBalance Asset       `json:"rex_balance"`
+	MaturedRex uint64      `json:"matured_rex"`
+
+	// TODO: set the exact type
+	RexMaturities []interface{} `json:"rex_maturities"`
+}
+
+// added 2022-11-02
```
Comment (maoueh) [inline types.go]: Same for Go struct comment format

File: types.go [hunk 12]
```diff
@@ -656,9 +684,9 @@ type PermissionLevelWeight struct {
 
 type Authority struct {
 	Threshold uint32                  `json:"threshold"`
-	Keys      []KeyWeight             `json:"keys,omitempty"`
-	Accounts  []PermissionLevelWeight `json:"accounts,omitempty"`
-	Waits     []WaitWeight            `json:"waits,omitempty"`
+	Keys      []KeyWeight             `json:"keys"`
+	Accounts  []PermissionLevelWeight `json:"accounts"`
+	Waits     []WaitWeight            `json:"waits"`
```
Comment (maoueh) [inline types.go]: Reasoning for removing the `omitempty`?

File: types.go [hunk 13]
```diff
@@ -1247,6 +1278,185 @@ func (i *Float128) UnmarshalJSON(data []byte) error {
 	return nil
 }
 
+// added 2022-11-02
```
Comment (maoueh) [inline types.go]: No need for this for code you are adding that is unrelated to version of EOSIO/LEAP

File: p2p_test.go [hunk 14]
```diff
@@ -361,7 +361,7 @@ func TestBlockHeader_UnmarshalJSON(t *testing.T) {
 
 	for _, test := range tests {
 		t.Run(test.name, func(t *testing.T) {
-			baseBlockFormat := `{"action_mroot":"","transaction_mroot":"","confirmed":0,"previous":"","producer":"","schedule_version":0,"timestamp":"0001-01-01T00:00:00","header_extensions":null%s}`
```
Comment (maoueh) [inline p2p_test.go]: I would like to ensure that the older format is still supported. Either in this test somehow or in another test dedicated to it (that maybe is already existing, just want to be sure there is no lost here).

File: types.go [hunk 15]
```diff
@@ -80,9 +81,14 @@ type VoterInfo struct {
 	Proxy             AccountName   `json:"proxy"`
 	Producers         []AccountName `json:"producers"`
 	Staked            Int64         `json:"staked"`
-	LastVoteWeight    Float64       `json:"last_vote_weight"`
-	ProxiedVoteWeight Float64       `json:"proxied_vote_weight"`
+	LastVoteWeight    Decimal       `json:"last_vote_weight"`
+	ProxiedVoteWeight Decimal       `json:"proxied_vote_weight"`
```
Comment (maoueh) [inline types.go]: I've review the `Decimal` type below, I'm quite unsure about this change. If the API decided to use a `Float64`, then probably that precision was not of the utmost importance, because float are know to be imprecise. They would have used a `Int` directly with an amount of decimals to represent precise number.

It will be also a breaking change for users because they will not be able to cast to `float64` directly and it seems no conversion exists. I'm not against breaking changes but in this case, I do not see the reason to do it.

File: types.go [hunk 16]
```diff
@@ -1247,6 +1278,185 @@ func (i *Float128) UnmarshalJSON(data []byte) error {
...
 }
 
+// added 2022-11-02
+
+// Int
+// Experimental implementation
+
+// used big.Int
+// experimental type, and will replace or be replaced with previous Int implementation
+type Int struct {
```
Comment (maoueh) [inline types.go]: Quite unsure about this new type. I see it's used for `Decimal` only right now, and I'm against the change from `Float64` to `Decimal` higher in the PR. So for now, I would leave this aside until we see good reason to have them.

At which point, using an existing 256 bits integer library seems more appropriate, like https://github.com/holiman/uint256/blob/master/uint256.go which despite it's name supported Int and Uint. This is what `go-ethereum` use for their 256 bits math.

**Review by maoueh — CHANGES_REQUESTED** [review]

File: responses.go (hunk 4 above)
Comment (psy2848048) [reply responses.go]: Appreciate it :)

File: types.go (hunk 15 above)
Comment (psy2848048) [reply types.go]: Yes, even I don't want to make ambreaking change, but want to ask your opinion.
You implemented big number presentation yourself but now native-like `big.Int` is broadly used.
Between replacing to big.Int or keeping your implmentation, which do you prefer?

Plus, in JSON assertion, wish not to compare the number but impossible
Let me try manual typing one by one

File: types.go (hunk 16 above)
Comment (psy2848048) [reply types.go]: Not intend to replace my types from the entire code, but it's about modifying the inside of the existing `Int/Float` type
About the above, want to hear your thought - whether you prefer big.Int or not.

If yes, try to modify the current Int(might be?)
If not, roll back to current Int and Float

File: p2p_test.go (hunk 14 above)
Comment (psy2848048) [reply p2p_test.go:364]: Then would it be possible that the response of the early block and recent block are different?

The responses of the mock server were captured from a couple of days ago.

File: testdata/mock_server/server.go (hunk 6 above)
Comment (psy2848048) [reply testdata/mock_server/server.go]: I always do in Python but no knowledge in Go
Thx very much :)

File: responses.go (hunk 8 above)
Comment (psy2848048) [reply responses.go]: https://github.com/psy2848048/eos-go/blob/b489eeb186b3aa46096c1deff3f75ed651d6ebf8/testdata/mock_server/chain_get_account.json#L238

File: responses.go (hunk 7 above)
Comment (psy2848048) [reply responses.go:232]: https://github.com/psy2848048/eos-go/blob/b489eeb186b3aa46096c1deff3f75ed651d6ebf8/testdata/mock_server/chain_get_account.json#L6

File: types.go (hunk 12 above)
Comment (psy2848048) [reply types.go]: https://github.com/psy2848048/eos-go/blob/b489eeb186b3aa46096c1deff3f75ed651d6ebf8/testdata/mock_server/chain_get_account.json#L35

File: responses.go (hunk 3 above)
Comment (psy2848048) [reply responses.go]: Glad to hear you are active `git blame` user :)
Tried to find when it was changed but the docs shows from EOSIO v2.0 and the specs are already changed
Let me state as `changed from EOSIO/Leap v2.0`

File: api_test.go (hunk 1 above)
Comment (psy2848048) [reply api_test.go]: https://github.com/eoscanada/eos-go/pull/191/commits/18d78b47dd7be8c506ce92f5422a4273eb0d1820

File: api_test.go (hunk 2 above)
Comment (psy2848048) [reply api_test.go]: https://github.com/eoscanada/eos-go/pull/191/commits/18d78b47dd7be8c506ce92f5422a4273eb0d1820

File: responses.go (hunk 3 above)
Comment (psy2848048) [reply responses.go]: https://github.com/eoscanada/eos-go/pull/191/commits/18d78b47dd7be8c506ce92f5422a4273eb0d1820

File: responses.go (hunk 7 above)
Comment (psy2848048) [reply responses.go]: https://github.com/eoscanada/eos-go/pull/191/commits/18d78b47dd7be8c506ce92f5422a4273eb0d1820

Comment (psy2848048) [reply responses.go]: https://github.com/psy2848048/eos-go/blob/b489eeb186b3aa46096c1deff3f75ed651d6ebf8/testdata/mock_server/chain_get_account.json#L6

File: types.go (hunk 10 above)
Comment (psy2848048) [reply types.go]: https://github.com/eoscanada/eos-go/pull/191/commits/18d78b47dd7be8c506ce92f5422a4273eb0d1820

File: types.go (hunk 11 above)
Comment (psy2848048) [reply types.go]: https://github.com/eoscanada/eos-go/pull/191/commits/18d78b47dd7be8c506ce92f5422a4273eb0d1820

File: types.go (hunk 13 above)
Comment (psy2848048) [reply types.go]: https://github.com/eoscanada/eos-go/pull/191/commits/18d78b47dd7be8c506ce92f5422a4273eb0d1820

File: responses.go (hunk 9 above)
Comment (psy2848048) [reply responses.go]: https://github.com/eoscanada/eos-go/pull/191/commits/76ee1c81a195b10d296659067a9f077657ea1cc4

File: testdata/mock_server/server.go (hunk 6 above)
Comment (psy2848048) [reply testdata/mock_server/server.go]: https://github.com/eoscanada/eos-go/pull/191/commits/af34b235a84921ad8c77541d4edc9671062da6b6
https://github.com/eoscanada/eos-go/pull/191/commits/24f21ffc3aaf77228c136de56d10f2e62d8a165e

File: p2p_test.go (hunk 14 above)
Comment (psy2848048) [reply p2p_test.go:364]: Would it be possible to define a version setter in the package root as below:

```go
type Version int

const (
    V1 EosioVersion = 1 + iota
    V2
)

var EosioVersion Version = V2

func SetVersion(ver Version) {
    EosioVersion = ver
}
```

and implement as below?:

```go
const blockTimestampFormat = "2006-01-02T15:04:05.999"

func (t BlockTimestamp) MarshalJSON() ([]byte, error) {
	strTime := t.Format(blockTimestampFormat)
	if len(strTime) == len("2006-01-02T15:04:05.5") {
		strTime += "00"
	} else if len(strTime) == len("2006-01-02T15:04:05") && EosioVersion == V2 {
		// modified since EOSIO/Leap v2.0
		// force formatting into YYYY-MM-DDTHH:MM:SS.sss
		strTime += ".000"
	}

	return []byte(`"` + strTime + `"`), nil
}
```

Comment (maoueh) [reply p2p_test.go]: @psy2848048?

File: types.go (hunk 16 above)
Comment (maoueh) [reply types.go]: My reasoning here is to align with whatever the canonical source code EOSIO/Leap platform uses. If it uses a `long`, then we should use a `int64`, if it uses `unsigned long`, we use `uint64`, etc. I don't see any reason to use `big.Int` when the EOSIO uses a fixed size type.

In the case of `Decimal`, I did not see the need for it yet. I've checked the system contract related to `VoterInfo` that you modified above to use `Decimal`, in the system contract, the value used there is a `double` (https://github.com/AntelopeIO/reference-contracts/blob/main/contracts/eosio.system/include/eosio.system/eosio.system.hpp#L302) so I don't see why using `Float64` would be wrong and what would be the gain of using `Decimal` here.

If there is something wrong with `Float64` in respect to for example how it serialize/deserialize according to EOSIO/Leap rules, I'm open to modify it.

File: types.go (hunk 12 above)
Comment (maoueh) [reply types.go]: Ok but we are still not aligned with the response since `null` is not `[]` (but to achieve it, we need to always instantiate slices to an empty slice, and the Go JSON serializer as no option to serialize `nil` slices to `[]`).

I'm kind of hesitant to change it although I'm pretty sure it has normally no impacts. 

Did the change help make tests pass?

File: p2p_test.go (hunk 14 above)
Comment (psy2848048) [reply p2p_test.go]: Sorry for confuse my previous comment which is deleted

Even I think that asserting "0001-01-01T00:00:00" and "0001-01-01T00:00:00.000" are not important, nor does "0" and "0.000000" but current `assert.JSONEq` cannot make it with tolerance. It seems that strict unmarshaling is not really productive if it is not related signing.

So I decide to delete a modified `BlockTimestamp` and `Int` & `Decimal`, and now working on implementing a kind of loosen JSONEq implementation - detect a real type from the string attribute, and convert into time and compare by Go time object if it types `Time`, comparing with rounded number if it types `Float`, and more

File: types.go (hunk 15 above)
Comment (psy2848048) [reply types.go]: By https://github.com/eoscanada/eos-go/pull/191#discussion_r1016782712 , the the type will be removed

Comment (maoueh) [reply types.go]: > You implemented big number presentation yourself but now native-like big.Int is broadly used

Not sure what you are referring to exactly, of which type you are talking about exactly?

> Plus, in JSON assertion, wish not to compare the number but impossible

Agree, it's harder to deal with full JSON assertion for some numbers. I would not go manual typing neither, maybe just change the JSON value to be 0.0 or something that always equivalent. The goal is to ensure the representation of the response is good, not that much that it's precise.

And like mentioned below, reference contracts uses `double` (https://github.com/AntelopeIO/reference-contracts/blob/main/contracts/eosio.system/include/eosio.system/eosio.system.hpp#L302) so I don't see a good reason to switch the type here, `float64` in Golang is a `double` in C++.

Please revert the type change and remove `Decimal/Int` types for now until there is a use for them.

File: types.go (hunk 16 above)
Comment (psy2848048) [reply types.go]: By https://github.com/eoscanada/eos-go/pull/191#discussion_r1016782712 , the the type will be removed

File: types.go (hunk 12 above)
Comment (psy2848048) [reply types.go]: By https://github.com/eoscanada/eos-go/pull/191#discussion_r1016782712 , let me try to contain to assert as same between nil and empty array

File: types.go (hunk 15 above)
Comment (psy2848048) [reply types.go]: By https://github.com/eoscanada/eos-go/pull/191#discussion_r1016782712 , the type will be removed

File: p2p_test.go (hunk 14 above)
Comment (maoueh) [reply p2p_test.go]: I would just like to ensure we are accepting both form for unmarshalling. I know in the past that with and without `.000` bit us.

For signing purposes, data is packed using a deterministic binary encoding, so JSON representation is not important for signing in EOSIO world.

About the loose `JSONEq`, I think it's not something that should be pursue. Detecting real type from string is always hard. What other ways could we have to make the tests pass without implementing a specialized `JSONEq`? Maybe just changing a bit the actual input?

File: types.go (hunk 16 above)
Comment (psy2848048) [reply types.go]: https://github.com/eoscanada/eos-go/pull/191/commits/ca5b8435b0ca0323c521abc3d07eb6985015a0f6

File: types.go (hunk 15 above)
Comment (psy2848048) [reply types.go]: https://github.com/eoscanada/eos-go/pull/191/commits/ca5b8435b0ca0323c521abc3d07eb6985015a0f6

File: p2p_test.go (hunk 14 above)
Comment (psy2848048) [reply p2p_test.go]: Appreciated and manipulated the mok expected result if the diff of the actual and expected result are almost same and meaningless

https://github.com/eoscanada/eos-go/pull/191/commits/ca5b8435b0ca0323c521abc3d07eb6985015a0f6

File: types.go (hunk 12 above)
Comment (psy2848048) [reply types.go]: https://github.com/eoscanada/eos-go/pull/191/commits/ca5b8435b0ca0323c521abc3d07eb6985015a0f6

File: responses.go (hunk 8 above)
Comment (psy2848048) [reply responses.go]: Rollbacked as an actual object and manipulated the expected result

https://github.com/eoscanada/eos-go/pull/191/commits/ca5b8435b0ca0323c521abc3d07eb6985015a0f6

File: types.go [hunk 17]
```diff
@@ -972,7 +994,7 @@ func (f *Float64) MarshalJSON() ([]byte, error) {
 		return []byte("\"nan\""), nil
 	default:
 	}
-	return json.Marshal(float64(*f))
+	return json.Marshal(fmt.Sprintf("%f", float64(*f)))
```
Comment (psy2848048) [inline types.go]: One big point to propose my another `Float` and the problem(assertion in different types) could be solved easier than I thought, and I thought that this change would have no problem in backward compatiblity because it changes only marshaling logic and no CI test collision from other modules. What do you think about this line change?

As-was
```
Diff:
                                --- Expected
                                +++ Actual
                                @@ -243,3 +243,3 @@
                                   (string) (len=8) "is_proxy": (float64) 0,
                                -  (string) (len=16) "last_vote_weight": (string) (len=19) "697460718699.945435",
                                +  (string) (len=16) "last_vote_weight": (float64) 6.974607186999454e+11,
                                   (string) (len=5) "owner": (string) (len=12) "teamgreymass",
                                @@ -247,3 +247,3 @@
                                   },
                                -  (string) (len=19) "proxied_vote_weight": (string) (len=8) "0.000000",
                                +  (string) (len=19) "proxied_vote_weight": (float64) 0,
```

File: responses.go (hunk 4 above)
Comment (psy2848048) [reply responses.go]: Checked & updated according to Leap repo
https://github.com/eoscanada/eos-go/pull/191/commits/9794c6ca7371c08add27485713eab9ad3fb6de3a

File: types.go (hunk 16 above)
Comment (psy2848048) [reply types.go]: @maoueh Any comment of them?

File: types.go (hunk 17 above)
Comment (maoueh) [reply types.go]: Seems like even worse breaking change IMO (although I like the general ideal), because now people marshalling will "silently" receive a `string` type instead of a `number` type in JSON.

Meaning you upgrade `eos-go` and marshalling type is different meaning consuming code needs to be updated (or you need hardest work to come back to it.

I have an in-between solution although it's imperfect. We could have a global variable that controls this (string vs number marshalling) and by default would stay with the old behavior but in the tests we would override it to print to string.

So, we have 3 possibilities here:
- Status quo (tests uses fixed value that are always precise)
- Override variable for testing purposes
- Breaking change `number` -> `string` but highly documented

Comment (psy2848048) [reply types.go]: I added a global value  the overriding plan here's my push
https://github.com/eoscanada/eos-go/pull/191/commits/ac0a8d441e670b415cb56963113f4eb1cd91a5ac

**Review by maoueh — APPROVED** [review]
Anything else to review on my side?

**Comment by psy2848048** [comment]
Added https://github.com/eoscanada/eos-go/pull/191/commits/23fe966b6359917c845793dbe1d3fd54c054c74b against Ci failure but cannot understand the fail 'cause `go test` includes `go mod tidy`

**Comment by psy2848048** [comment]
Bumped up other indirect dependencies related with gjson, as [496c90c](https://github.com/eoscanada/eos-go/pull/191/commits/496c90cc38834223ff51cee41cb55381b87f03e5)
